src/wallet/walletdb.cpp

```diff
@@ -855,6 +751,263 @@ static DBErrors LoadWalletFlags(CWallet* pwallet, DatabaseBatch& batch) EXCLUSIV
     return DBErrors::LOAD_OK;
 }
 
+struct LoadResult
+{
+    DBErrors m_result{DBErrors::LOAD_OK};
+    std::string m_error{};
+    int m_records{0};
+};
+
+using LoadFunc = std::function<DBErrors(CWallet* pwallet, DataStream& key, CDataStream& value, std::string& err)>;
+static LoadResult LoadRecords(CWallet* pwallet, DatabaseBatch& batch, const std::string& key, LoadFunc load_func)
+{
+    LoadResult result;
+    DataStream ssKey;
+    CDataStream ssValue(SER_DISK, CLIENT_VERSION);
+
+    DataStream prefix;
+    prefix << key;
+    std::unique_ptr<DatabaseCursor> cursor = batch.GetNewPrefixCursor(prefix);
+    if (!cursor) {
+        pwallet->WalletLogPrintf("Error getting database cursor for '%s' records\n", key);
+        result.m_result = DBErrors::CORRUPT;
+        return result;
+    }
+
+    while (true) {
+        DatabaseCursor::Status status = cursor->Next(ssKey, ssValue);
+        if (status == DatabaseCursor::Status::DONE) {
+            break;
+        } else if (status == DatabaseCursor::Status::FAIL) {
+            pwallet->WalletLogPrintf("Error reading next '%s' record for wallet database\n", key);
+            result.m_result = DBErrors::CORRUPT;
+            return result;
+        }
+        std::string type;
+        ssKey >> type;
+        assert(type == key);
+        if ((result.m_result = std::max(result.m_result, load_func(pwallet, ssKey, ssValue, result.m_error))) == DBErrors::CORRUPT) {
+            pwallet->WalletLogPrintf("%s\n", result.m_error);
+        }
+        ++result.m_records;
+    }
+    return result;
+}
+
+static DBErrors LoadLegacyWalletRecords(CWallet* pwallet, DatabaseBatch& batch, int last_client) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
+{
+    AssertLockHeld(pwallet->cs_wallet);
+    DBErrors result = DBErrors::LOAD_OK;
+
+    // Make sure descriptor wallets don't have any legacy records
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+        for (const auto& type : DBKeys::LEGACY_TYPES) {
+            DataStream key;
+            CDataStream value(SER_DISK, CLIENT_VERSION);
+
+            DataStream prefix;
+            prefix << type;
+            std::unique_ptr<DatabaseCursor> cursor = batch.GetNewPrefixCursor(prefix);
+            if (!cursor) {
+                pwallet->WalletLogPrintf("Error getting database cursor for '%s' records\n", type);
+                return DBErrors::CORRUPT;
+            }
+
+            DatabaseCursor::Status status = cursor->Next(value, value);
```

Comment:
In 0506cc23:
The key nor the value are used here but this should be `(key, value)` instead of `(value, value)`.

Reply:
oops, fixed